mmhuman3d/core/visualization/renderer/torch3d_renderer/base_renderer.py

```diff
@@ -163,147 +226,51 @@ def __del__(self):
-        return cameras
-
-    def init_renderer(self, cameras, lights):
-        """Initial renderer."""
-        raster = build_raster(
-            dict(
-                type=self.raster_type,
-                cameras=cameras,
-                raster_settings=self.raster_settings))
-        renderer = MeshRenderer(
-            rasterizer=raster,
-            shader=build_shader(
-                dict(
-                    type=self.shader_type,
-                    cameras=cameras,
-                    device=self.device,
-                    lights=lights,
-                    materials=self.materials,
-                    blend_params=self.blend_params))
-            if self.shader_type != 'silhouette' else build_shader(
-                dict(type=self.shader_type, blend_params=self.blend_params)))
-        return renderer
-
-    @staticmethod
-    def rgb2bgr(rgbs):
-        if isinstance(rgbs, torch.Tensor):
-            bgrs = torch.cat(
-                [rgbs[..., 0, None], rgbs[..., 1, None], rgbs[..., 2, None]],
-                -1)
-        elif isinstance(rgbs, np.ndarray):
-            bgrs = np.concatenate(
-                [rgbs[..., 0, None], rgbs[..., 1, None], rgbs[..., 2, None]],
-                -1)
-        return bgrs
-
-    @staticmethod
-    def image_tensor2numpy(image):
-        return (image.detach().cpu().numpy() * 255).astype(np.uint8)
-
-    def write_images(self, rgbs, valid_masks, images, indexes):
+    def _write_images(self, rgba, backgrounds, indexes):
         """Write output/temp images."""
-        rgbs = rgbs.clone()[..., :3] / rgbs[..., :3].max()
-        bgrs = torch.cat(
-            [rgbs[..., 0, None], rgbs[..., 1, None], rgbs[..., 2, None]], -1)
-        if images is not None:
-            output_images = bgrs * 255 * valid_masks + (1 -
-                                                        valid_masks) * images
-            output_images = output_images.detach().cpu().numpy().astype(
-                np.uint8)
+        if rgba.shape[-1] > 3:
+            rgbs, valid_masks = rgba[..., :3], rgba[..., 3:]
         else:
-            output_images = (bgrs.detach().cpu().numpy() * 255).astype(
-                np.uint8)
+            rgbs = rgba[..., :3]
+            valid_masks = torch.ones_like(rgbs)[..., :1]
+        rgbs = normalize(rgbs, origin_value_range=(0, 1), clip=True)
+        bgrs = rgb2bgr(rgbs)
+        if backgrounds is not None:
+            image_max = 1.0 if backgrounds.max() <= 1.0 else 255
+            backgrounds = normalize(
+                backgrounds,
+                origin_value_range=(0, image_max),
+                out_value_range=(0, 1))
+            output_images = bgrs * valid_masks + (1 -
+                                                  valid_masks) * backgrounds
+            output_images = tensor2array(output_images)
+
+        else:
+            output_images = tensor2array(bgrs)
         for idx, real_idx in enumerate(indexes):
```

Comment:
Why not using `array_to_images`.

Reply:
I'm not sure whether `array_to_images` is reliable. Need check